tests/test_sort_pupilimg_frames.py

```diff
@@ -0,0 +1,569 @@
+if np.all(np.sign(np.diff(EXPTIME_KGAIN[:-1])) == 1) is False:
+    raise Exception('Exposure times in K-gain must be monotonic but for the last value')
+
+# Non-linearity
+EXPTIME_NONLIN = [0.076, 0.758, 1.515, 2.273, 3.031, 3.789, 4.546, 5.304, 6.062,
+    6.820, 7.577, 8.335, 9.093, 9.851, 10.608, 11.366, 12.124, 12.881, 13.639,
+    14.397, 15.155, 1.515]
+CMDGAIN_NONLIN = [1.65, 5.24, 8.60, 16.70, 27.50, 45.26, 87.50, 144.10, 237.26,
+    458.70, 584.40]
+# Checks
+if len(EXPTIME_NONLIN) < 22:
+    raise Exception(f'Insufficient frames ({len(EXPTIME_NONLIN)}) per unique EM value in non-linearity')
+if len(CMDGAIN_NONLIN) < 11:
+    raise Exception(f'Insufficient values of distinct EM Values ({len(EXPTIME_NONLIN)}) in non-linearity')
+if np.sum(np.array(EXPTIME_NONLIN) == EXPTIME_NONLIN[-1]) != 2:
+    raise Exception('Last exposure time must be repeated once')
+if len(set(EXPTIME_NONLIN)) != len(EXPTIME_NONLIN) - 1:
+    raise Exception('Only one exposure time can be repeated in non-linearity')
+if EXPTIME_NONLIN[-1] in EXPTIME_NONLIN[0:5] is False:
+    raise Exception('The last exposure time must be present at the beginning of the exposure times in non-linearity')
+if np.all(np.sign(np.diff(EXPTIME_NONLIN[:-1])) == 1) is False:
+    raise Exception('Exposure times in Non-linearity must be monotonic but for the last value')
+
+# Notice the pairing between unity and non-unity gain frames
+EM_EMGAIN=[1.000, 1.000, 1.007, 1.015, 1.024, 1.035, 1.047, 1.060, 1.076, 1.094,
+    1.115, 1.138, 1.165, 1.197, 1.234, 1.276, 1.325, 1.385, 1.453, 1.534, 1.633,
+    1.749, 1.890, 2.066, 2.278, 2.541, 2.873, 3.308, 3.858, 4.581, 5.577, 6.189,
+    6.906, 7.753, 8.757, 9.955, 11.392, 13.222, 15.351, 17.953, 21.157, 25.128,
+    30.082, 36.305, 44.621, 54.768, 67.779, 84.572, 106.378, 134.858, 172.244,
+    224.385, 290.538, 378.283, 494.762, 649.232, 853.428]
+EXPTIME_EMGAIN_SEC=[5, 10, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
+    5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 10, 10, 10,
+    10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
+NFRAMES_EMGAIN=[3, 5, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+    3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 5, 5, 5,
+    5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5]
+# Checks
+if len(EM_EMGAIN) < 56:
+    raise Exception(f'Insufficient number of EM gain values ({len(EM_EMGAIN)}) in EM-gain vs DAC')
+if len(EXPTIME_EMGAIN_SEC) != len(EM_EMGAIN):
+    raise Exception(f'Inconsistent number of sets in EM-gain vs DAC')
+if len(EXPTIME_EMGAIN_SEC) != len(NFRAMES_EMGAIN):
+    raise Exception(f'Inconsistent number of sets in EM-gain vs DAC')
+
+# Test data: Consider two possible scenarios: identical exposure times among the
+# different subsets of non-unity gain used to calibrate non-linearity or different
+# values of exposure times
+
+# Values for mean frame
+cmdgain_mean_frame, exptime_mean_frame = get_cmdgain_exptime_mean_frame(
```

Comment:
In general, if you have this much setup code, it's good practice to put it in a setup function. pytest has a `setup_module()`  function that you can define. this runs setup code for a .py file. I would put all this code in there, and use the python `global` declaration inside that function to identify the variables you want to grab for the unit tests. 

```
def setup_module():
    global dataset_w_change, dataset_wo_change # this will make these variables you create in this function accessible to other functions
    # your code here
```

Reply:
Indeed I tried myself, because that amount of block code is quite ugly as-is, though I did not know the precise syntax of the function

Reply:
Done